Lib/test/test_asyncio/test_base_events.py

```diff
@@ -861,20 +861,15 @@ async def raise_keyboard_interrupt():
 
         self.loop._process_events = mock.Mock()
 
-        try:
+        with self.assertRaises(KeyboardInterrupt):
             self.loop.run_until_complete(raise_keyboard_interrupt())
-        except KeyboardInterrupt:
-            pass
 
         def func():
             self.loop.stop()
             func.called = True
         func.called = False
-        try:
-            self.loop.call_soon(func)
-            self.loop.run_forever()
-        except KeyboardInterrupt:
-            pass
+        self.loop.call_later(0.01, func)
```

Comment:
A bit unfortunate that we had to add a 10 msec delay here -- in most cases that will waste 10 msec and occasionally it will not wait long enough and the test will flake-fail. Why won't call_soon work?

Reply:
Background: this test covers a bug that base exceptions like `KeyboardInterrupt` will cause the loop to be closed twice, leading to an early exit in the next loop run.

This test was added before `_StopError` was dropped, at that time the loop will stop as soon as `_StopError` is raised, even if the `_ready` queue still has a handle (which is the next run - the `func()` here), therefore this was a good test. But now the loop will exhaust the `_ready` queue before stopping, so this test couldn't cover the bug anymore without this PR. That's also why we cannot simply add `func()` into the `_ready` queue by `call_soon()`.

You're right about the flakyness of using `call_later()` tho. Now I think a better fix would be to use chained `call_soon()` to skip the first iteration where the buggy stop occured:

```suggestion
        # If the issue persists, the loop will exit early after the first iteration,
        # in that case our func() in the second iteration won't be called.
        self.loop.call_soon(self.loop.call_soon, func)
```

I verified and this error fails if the fix in f3e2e09213ecac36c3c3634a06cb7badf3e4217e is reverted.

Reply:
Clever solution, go ahead and make it a PR, assign to me or @kumaraditya303 for review.